algorithms/a3c/main.py

```diff
@@ -40,80 +54,276 @@
+            idx_of_latest = checkpoint_file_name_ints.index(max(checkpoint_file_name_ints))
+            checkpoint_to_load = checkpoint_paths[idx_of_latest]
+            print('Attempting to load latest checkpoint: {}'.format(checkpoint_to_load))
+
+            # load checkpoint and unpack values
+            if os.path.isfile(checkpoint_to_load):
+                print("Successfully loaded checkpoint {}".format(checkpoint_to_load))
+                checkpoint = torch.load(checkpoint_to_load)
+                args.total_length = checkpoint['total_length']
+                args.episode_number = checkpoint['episode_number']
+                args.num_backprops = checkpoint.get('num_backprops', 0)
+                checkpoint_counter = checkpoint.get('counter', False)  # if not set, set to 0 later
+
+                print('Values from checkpoint: total_length: {}. episode_number: {}'.format(
+                    checkpoint['total_length'], checkpoint['episode_number']))
+                shared_model.load_state_dict(checkpoint['state_dict'])
+                optimizer.load_state_dict(checkpoint['optimizer'])
+
+                print("=> loaded checkpoint '{}' (total_length {})"
+                      .format(checkpoint_to_load, checkpoint['total_length']))
+        else:
+            print('No model checkpoint to load')
 
-    if not args.synchronous:
-        # test runs continuously and if episode ends, sleeps for args.test_sleep_time seconds
-        p = mp.Process(target=test, args=(args.num_processes, args, shared_model, counter))
-        p.start()
-        processes.append(p)
+    # Creating TensorBoard writer and necessary folders
+    writer = SummaryWriter(comment='A3C',  # this will create dirs
+                           log_dir=args.tensorboard_path, purge_step=args.episode_number)
+    # run tensorboardX --logs_dir args.tensorboard_path in terminal and open browser e.g.
+    print('-----------------\nTensorboard command:\n'
+          'tensorboard --logdir experiments/{}/tensorboard_logs'
+          '\n-----------------'.format(args.experiment_id))
 
-        for rank in range(0, args.num_processes):
-            p = mp.Process(target=train, args=(rank, args, shared_model, counter, lock, optimizer))
+    # Save argparse arguments and environment config from last resume or first start
+    with open(os.path.join(args.experiment_path, 'latest_args.json'), 'w') as f:
+        args_dict = vars(args)
+        args_dict['experiment_id'] = str(args.experiment_id)
+        json.dump(args_dict, f)
+    if not args.vizdoom and not args.atari:
+        # Save ai2thor config
+        with open(args.last_config_resume_path, 'w') as f:
+            json.dump(env.config, f)
+
+    # todo try fix multiprocessing again
+    # process initialisation and training/testing starting
+    processes = []
+    counter = mp.Value('i', 0 if not checkpoint_counter else checkpoint_counter)
+    lock = mp.Lock()
+    try:
+        if not args.synchronous:
```

Comment:
pep8: if args.synchronous what you put in the else, else what you put in this if. Always formulate positive ifs if you can